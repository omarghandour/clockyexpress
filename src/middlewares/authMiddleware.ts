import jwt from "jsonwebtoken";
import { Request, Response, NextFunction } from "express";
import User from "../models/user";

interface AuthRequest extends Request {
  user?: any;
}

const authMiddleware = async (
  req: AuthRequest,
  res: Response,
  next: NextFunction
) => {
  let token;

  if (
    req.headers.authorization &&
    req.headers.authorization.startsWith("Bearer")
  ) {
    try {
      token = req.headers.authorization.split(" ")[1];
      const decoded = jwt.verify(token, process.env.JWT_SECRET!);
      req.user = await User.findById((decoded as any).id).select("-password");

      // Check if the user is an admin
      if (req.user && req.user.isAdmin) {
        console.log("User is admin");

        next();
      } else {
        res.status(403).json({ message: "Access denied. Admins only." });
      }
    } catch (error) {
      console.log(error);

      res.status(401).json({ message: "Not authorized, token failed" });
    }
  }

  if (!token) {
    console.log("no token");

    res.status(401).json({ message: "Not authorized, no token" });
  }
};

export { authMiddleware };
