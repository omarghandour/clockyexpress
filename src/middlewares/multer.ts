import multer from "multer";
import path from "path";

// Set storage engine for multer
const storage = multer.diskStorage({
  destination: function (req, file, cb) {
    cb(null, "uploads/"); // Directory where files will be stored
  },
  filename: function (req, file, cb) {
    cb(null, `${Date.now()}-${file.originalname}`); // Unique file name
  },
});

// File filter to check if the uploaded file is an image
const fileFilter = (req: any, file: any, cb: any) => {
  const allowedTypes = ["image/jpeg", "image/jpg", "image/png"];
  if (allowedTypes.includes(file.mimetype)) {
    cb(null, true);
  } else {
    cb(new Error("Invalid file type. Only JPEG, JPG, and PNG are allowed."));
  }
};

// Initialize multer
const upload = multer({
  storage: storage,
  fileFilter: fileFilter,
  limits: { fileSize: 1024 * 1024 * 5 }, // 5 MB size limit
});

export default upload;
