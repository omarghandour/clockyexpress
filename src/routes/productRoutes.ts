import { Router } from "express";
import {
  getProducts,
  addProduct,
  removeProduct,
  getProductById,
} from "../controllers/productController";
import { authMiddleware } from "../middlewares/authMiddleware";

const router = Router();

router.route("/").get(getProducts).post(authMiddleware, addProduct);
router.route("/:PID").get(getProductById).delete(authMiddleware, removeProduct);
export default router;
