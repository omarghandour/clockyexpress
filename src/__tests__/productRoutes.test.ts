import request from "supertest";
import express from "express";
import productRoutes from "../routes/productRoutes";
import Product from "../models/product";
import { authMiddleware } from "../middlewares/authMiddleware";
jest.mock("../middlewares/authMiddleware", () => ({
  authMiddleware: jest.fn(),
}));
// Mock the Product model
jest.mock("../models/product");

const app = express();
app.use(express.json());
app.use("/api/products", productRoutes);

describe("Product Routes", () => {
  it("should return a list of products and match snapshot", async () => {
    const products = [
      {
        name: "Product 1",
        price: 100,
        description: "Test product",
        countInStock: 10,
      },
      {
        name: "Product 2",
        price: 200,
        description: "Test product 2",
        countInStock: 5,
      },
    ];

    (Product.find as jest.Mock).mockResolvedValue(products);

    const res = await request(app).get("/api/products");

    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });
  it("should return a Product by ID", async () => {
    const product = {
      _id: "123abc",
      name: "Product 1",
      price: 100,
      description: "Test product",
      countInStock: 10,
    };

    (Product.findById as jest.Mock).mockResolvedValue(product);

    const res = await request(app).get(`/api/products/${product._id}`);

    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });
  it("should delete a Product by ID", async () => {
    const product = {
      _id: "123abc",
      name: "Product 1",
      price: 100,
      description: "Test product",
      countInStock: 10,
    };

    (Product.findByIdAndDelete as jest.Mock).mockResolvedValue(product);
    (authMiddleware as jest.Mock).mockImplementation((req, res, next) => {
      req.user = { _id: "123", isAdmin: true }; // Simulate an admin user
      next();
    });

    const res = await request(app)
      .delete(`/api/products/${product._id}`)
      .set("Authorization", "Bearer fakeToken"); // Simulate a token being sent

    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });
  it("should create a new product with valid token", async () => {
    const product = {
      _id: "1",
      name: "Sample Product",
      price: 99.99,
      description: "This is a sample product",
      countInStock: 10,
    };

    (Product.create as jest.Mock).mockResolvedValue(product);

    // Mock the implementation of authMiddleware
    (authMiddleware as jest.Mock).mockImplementation((req, res, next) => {
      req.user = { _id: "123", isAdmin: true }; // Simulate an admin user
      next();
    });

    const res = await request(app)
      .post("/api/products")
      .set("Authorization", "Bearer fakeToken") // Simulate a token being sent
      .send({
        name: "Sample Product",
        price: 99.99,
        description: "This is a sample product",
        countInStock: 10,
      });

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot(); // Compare the response with the snapshot
  });

  it("should return 401 if no token is provided", async () => {
    (authMiddleware as jest.Mock).mockImplementation((req, res, next) => {
      res.status(401).json({ message: "Not authorized, no token" });
    });

    const res = await request(app).post("/api/products").send({
      name: "Sample Product",
      price: 99.99,
      description: "This is a sample product",
      countInStock: 10,
    });

    expect(res.status).toBe(401);
    expect(res.body).toEqual({ message: "Not authorized, no token" });
  });
  // Add more tests here
});
