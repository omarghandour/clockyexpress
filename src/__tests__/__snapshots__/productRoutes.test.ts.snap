// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Routes should create a new product with valid token 1`] = `""`;

exports[`Product Routes should delete a Product by ID 1`] = `
{
  "message": "Product deleted successfully",
}
`;

exports[`Product Routes should return a Product by ID 1`] = `
{
  "_id": "123abc",
  "countInStock": 10,
  "description": "Test product",
  "name": "Product 1",
  "price": 100,
}
`;

exports[`Product Routes should return a list of products and match snapshot 1`] = `
[
  {
    "countInStock": 10,
    "description": "Test product",
    "name": "Product 1",
    "price": 100,
  },
  {
    "countInStock": 5,
    "description": "Test product 2",
    "name": "Product 2",
    "price": 200,
  },
]
`;
